Controllers films users

rovniipoc opened the pull request:
Добрый день.
Представляю на ревью проект по ТЗ 10го спринта.

rovniipoc committed:
Реализовать FilmController и UserController

rovniipoc committed:
Реализовать FilmController и UserController

rovniipoc committed:
Реализовать валидация моделей через аннотации @NotNull и @NotBlank

rovniipoc committed:
Откатить валидацию моделей через аннотации @NotNull и @NotBlank

При валидации через аннотации сервер принимал и обрабатывал без ошибок null поля. Пришлось откатиться на валидацию в контроллерах

rovniipoc committed:
Реализовать валидацию моделей через аннотации @NotNull, @NotBlank, @Email, @Past, @Pattern

Проверено: работает (в прошлый раз забыл аннотацию @Valid)

rovniipoc committed:
Удалить лишнюю валидацию логина в UserController

rovniipoc committed:
Добавить FilmControllerTests и 2 теста

В FilmControllerTests добавлены тесты:
createFilmsAndGetAllFilmsShouldBeOk()
createEmptyNameFilmShouldBeFail()

rovniipoc committed:
Добавить 2 теста в FilmControllerTests, заменить методы проверки длины описания на аннотации

В FilmControllerTests добавлены тесты:
tooLongDescriptionFilmShouldBeFail()
tooOldReleaseDateFilmShouldBeFail()

rovniipoc committed:
Добавить все тесты в FilmControllerTests и UserControllerTests, заменить метод проверки продолжительности фильма на аннотацию, изменить метод обновления фильмов

rovniipoc committed:
Исправить в соответствии с код-стайлом

rovniipoc committed:
Исправить в соответствии с код-стайлом

rovniipoc committed:
Исправить в соответствии с код-стайлом

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/FilmorateApplication.java:
```diff
@@ -1,12 +1,17 @@
+        SpringApplication.run(FilmorateApplication.class, args);
 
+        while (true) {
+
```
хм)

YazykovMikhail replied:
Как будто бы лишнее

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,75 @@
 public class FilmController {
+
+    private final Map<Long, Film> films = new HashMap<>();
+    private static final LocalDate VALID_RELEASE_DATE = LocalDate.of(1895, 12, 28);
```
статические поля обьвляются первыми - по кодстайлу

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,75 @@
 
 @RestController
+@RequestMapping("/films")
+@Slf4j
```
Стандартный шаблон для логов
createMethod(Request request) {
log.info("пришел Post запрос /films с телом: {}", request)
Response response = //логика метода, если есть бизнес процессы для логирвоания логируем их
log.info("Отправлен ответ Post /films с телом: {}", response)
}
что бы было видно что пришло, что ушло и если были какие то изменения в процессы выполнянения тоже были видны, потому что ты пойдешь в логи когда у тебя что то сломалось, просто по сообщения "Фильм добалвен" ничего не понятно.

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,75 @@
+
+    @PostMapping
+    public Film create(@Valid @RequestBody Film film) {
+        if (film.getReleaseDate().isBefore(VALID_RELEASE_DATE)) {
```
методы дублируются(в update), вынеси в отдельный метод

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,75 @@
+            } else {
+                oldFilm.setReleaseDate(film.getReleaseDate());
+            }
+            oldFilm.setName(film.getName());
```
Просто полностью заменй обьект
films.put(film.getId(), film)

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,75 @@
+            if (film.getReleaseDate().isBefore(VALID_RELEASE_DATE)) {
+                log.trace("Дата релиза фильма \"{}\" изменена", oldFilm.getName());
+                throw new ValidationException("Дата релиза не может быть раньше " + VALID_RELEASE_DATE);
+            } else {
```
этот else лишний 
просто валидируй, а потом оновляй полностью обьект

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,75 @@
+        throw new NotFoundException("Id = " + film.getId() + " не найден");
+    }
+
+    private long getNextId() {
```
Представь что у тьея 97678й987 фильмов и тебе нужно добавить новый, приедтся пробежаться по всем и вычислить новое id  и так каждый раз
Сделай поле счетчик идентификатора как было в TaskManager и увеличивай его перед установкой значения

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,23 @@
 public class Film {
+    private Long id;
+    @NotNull
+    @NotBlank
```
@NotNull
    @NotBlank
Посмотри между ними разницу и оставь что-то одно

YazykovMikhail replied:
Поправь везде

YazykovMikhail reviewed (changes requested):
Комментарии касаются и User и Film

rovniipoc committed:
Доработать по результатам 1-го ревью

Изменения:
Добавлен плагин maven-checkstyle
Удалены аннотации @NotNull
Убраны лишние объекты oldUser, oldFilm
Изменен счетчик id
Добавлен метод releaseDateValidation()
Добавлен метод nameUpdate()
Добавлен метод emailValidate()
Исправлено логирование
Оптимизирована проверка id при обновлении film/user

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,65 @@
+    @GetMapping
+    public Collection<Film> findAll() {
+        log.info("Пришел запрос Get /films");
+        log.info("Отправлен ответ Get /films с телом {}", films.values());
```
1) Делаешь два вызова в хранилище
2) Данные еще не сформированы для ответа а ты делаешь уже лог что уже готово
3) Между этими азпросами данные могут поменяться и будет рассинхрон

Сохарни в локальную переменну, сделай лог и верни ответ

YazykovMikhail replied:
Поправь и для User
